Add dynamic page range slider for uploaded PDFs with validation

AdiBak opened the pull request:
## Summary

This PR introduces a dynamic page range slider for PDF uploads. It allows users to select a valid subset of pages to process and ensures that the selected range respects the actual length of the uploaded file.

## Why This Matters

Previously, there was no such way to control which pages were extracted from a PDF. This adds a user-friendly, intuitive way to extract specific pages, which can save time and focus on relevant content. It also handles out-of-bounds inputs gracefully.

## Implementation

- Used `st.session_state` to store:
  - Uploaded PDF file
  - Slider values (`start` and `end`)
- Introduced a `validate_page_range(start, end, n_pages)` helper method in `file_manager.py`:
  - Ensures `start >= 1`, `end <= n_pages`
  - If `start >= end`, falls back to `start = 1`

## Behavior

- Slider exists in the sidebar and is initially set to 1-50
- After a new PDF is uploaded, the user can modify the slider as needed and send the chat prompt 
- A short note then appears below the slider to indicate the page range and file name 
- If the same PDF is reused, the user can still control the slider and obtain results
- Ranges like `(16, 42)` on a 23-page file are gracefully corrected to `(16, 23)`, while ranges like `(25, 42)` on a 23-page file are corrected to `(1, 23)`. 

## Files Touched

- `app.py`                 
- `file_manager.py`  
- `strings.json`          (chat field placeholder)

Thank you for your support and feedback!

barun-saha commented:
Addresses #99.

barun-saha commented:
Hi,

Thanks for the PR. I'll have a detailed look later. Meanwhile, a few issues that should be addressed:
- Please use single quotes for strings in the Python code.
- Please format & indent multi-line code properly. Please refer to the existing code to understand the style.

barun-saha commented on app.py:
```diff
@@ -222,6 +223,11 @@ def reset_api_key():
             )
 
+        # make slider with initial values
+        page_range_slider = st.slider("7: Specify a page range:",
```
Can you please introduce a constant, `MAX_ALLOWED_PAGES`, in global config, set the value to 150, and use the constant here? Let's consider the possibility that users may upload PDF files with more pages and make the code cleaner.

barun-saha replied:
To be more clear, please make it "7: Specify a page range for the PDF file:"

barun-saha commented on helpers/file_manager.py:
```diff
@@ -19,22 +19,50 @@
 def get_pdf_contents(
         pdf_file: st.runtime.uploaded_file_manager.UploadedFile,
+        page_range: tuple[int, int],
         max_pages: int = GlobalConfig.MAX_PAGE_COUNT
```
The `max_pages` arg is no longer used. Please update the function signature and the docstring.

barun-saha commented on helpers/file_manager.py:
```diff
@@ -19,22 +19,50 @@
+
+    :param pdf_file: The uploaded PDF file.
+    :param start: The start page 
+    :param max_pages: The end page
```
The `max_pages` arg is no longer used. Please update the function signature and the docstring. Also, `end` is not documented.

barun-saha commented on helpers/file_manager.py:
```diff
@@ -19,22 +19,50 @@
 
-    for page in range(n_pages):
-        page = reader.pages[page]
+    start, end = page_range                # set start and end per the range (user-specified values)
```
As noted earlier in the PR, please fix the code formatting for all the changes. Inline comments begin two spaces after the code.

barun-saha commented on helpers/file_manager.py:
```diff
@@ -19,22 +19,50 @@
-        page = reader.pages[page]
+    start, end = page_range                # set start and end per the range (user-specified values)
+    
+    print(f"starting at {start}, ending {end}")
```
This can be removed now.

barun-saha commented on helpers/file_manager.py:
```diff
@@ -19,22 +19,50 @@
+    if start > end:  # if the start is higher than the end, make it 1
+        start = 1
+
+    return (start, end)
```
Please remove redundant parentheses and change it to `return start, end`.

barun-saha commented on strings.json:
```diff
@@ -35,6 +35,6 @@
         "Did you know that SlideDeck AI can create a presentation based on any uploaded PDF file?"
     ],
-    "chat_placeholder": "Write the topic or instructions here. You can also upload a PDF file.",
+    "chat_placeholder": "Click or drop to upload a PDF and write the topic or instructions here.",
```
SlideDeck AI works even without an input PDF file. I fear that this revised statement conveys a sense that a PDF file might be mandatory. So, can we please roll back this change and retain the original statement?

barun-saha replied:
This text has not been reverted back.

barun-saha reviewed (changes requested):
Hi,

In addition to my earlier comments, I have also added some code review comments. Please have a look and update the PR. Thanks.

AdiBak commented:
Hi,
Thank you for the feedback. I committed these changes accordingly:

- utilized single quotes rather than double for strings
- placed inline comments two spaces after code
- formatted multiline code properly, placing lines underneath each other like the existing code 
- removed the unused PdfReader import, the unused max_pages parameter, and the redundant parentheses 
- added `MAX_ALLOWED_PAGES=150` to `global_config.py` and used this at slider initialization 
- clarified the slider label
- restored the original chat placeholder text 
- removed the redundant `else` block, and placed the statements to validate the page range, add the sidebar label for 'Extracting pages __ in ___.pdf', and get the pdf contents, to be just outside the `if prompt['files']:` block (so the `if` simply assigns the session state for the PDF file). I realized doing this preserved the functionality and made the code more clean.

These modifications surely helped improve the cleanliness and consistency of the code. Please let me know if it looks good or any further refinement. Thanks for your time and input!

barun-saha commented on app.py:
```diff
@@ -282,11 +290,29 @@ def set_up_chat_ui():
+                                                                            )
+        # show sidebar text for page selection and file name
+        with st.sidebar:
+            st.text(f'Extracting pages {st.session_state['start_page']} to {st.session_state['end_page']} in {st.session_state['pdf_file'].name}')
```
This line is still wrong. One can't have single quotes inside single quotes. Use double quotes inside.

barun-saha commented on app.py:
```diff
@@ -282,11 +290,29 @@ def set_up_chat_ui():
-            print(f'{prompt["files"]=}')
 
+        # get validated page range 
+        st.session_state['start_page'], st.session_state['end_page'] = filem.validate_page_range(
```
The page range can be checked only if a PDF file is available. If no PDF file has been uploaded, a run-time error will be thrown.

barun-saha commented on app.py:
```diff
@@ -282,11 +290,29 @@ def set_up_chat_ui():
+                                                                                st.session_state['end_page']
+                                                                            )
+        # show sidebar text for page selection and file name
+        with st.sidebar:
```
Same here. Lines 300 through 314 should be under the preceding `if` block.

barun-saha reviewed (changes requested):
Still some more fixes are required. Please do test all aspects of the application after making the changes and verify that there is no error.

AdiBak commented:
Hi,

My apologies if this response is a bit late. I refactored the code to:

- use single quotes and format as appropriate
- revert the chat field's placeholder text
- ensure no error is thrown when no PDF is initially uploaded 
- allow single page to be extracted (unsure if this is something to allow)
   - single pages do not always seem to work, though – when a page range of length 1, or a 1-page PDF is selected, then the app seems to display content from the previous PDF along with the 1 page's text, even though printing the extracted text every time does show the correct text for the current PDF. 
   
Please let me know if it looks OK or if any possible improvements. Thank you so much for your time and feedback!

barun-saha commented:
Hi,

Thanks for your contributions. I'll accept the current changes and merge them.

Having said that:

- The code formatting is still bad in some places. I'll fix them later separately.
- The code introduces a bug, resulting in a runtime error, since the slider is initialized only for the online mode. I'll fix that as well, separately.
